stlab/concurrency/future.hpp

```diff
@@ -629,14 +629,17 @@ struct shared<R(Args...)> : shared_base<R>, shared_task<Args...> {
     void remove_promise() override {
         if (std::is_same<R, reduced_t<R>>::value) {
             if (--_promise_count == 0) {
-                std::unique_lock<std::mutex> lock(this->_mutex);
-                if (!this->_ready) {
-                    this->reset();
-                    _f = function_t();
-                    this->_exception =
-                        std::make_exception_ptr(future_error(future_error_codes::broken_promise));
-                    this->_ready = true;
+                auto was_not_ready{false};
+                {
+                    std::unique_lock<std::mutex> lock(this->_mutex);
+                    if (!this->_ready) {
+                        was_not_ready = true;
+                        _f = function_t();
```

Comment:
I don't like destructing user-defined data under the lock. I believe that it is safe to access _f as no other promise can have access so no need to test _ready (tradeoff here, _ready is atomic), _f is a pointer comparison). I don't think we need to explicitly destruct _f as it should go away as soon as the shared state is dropped (in the destructor, calling remove_promise()). So this collapses to:

```cpp
if ((--_promise_count == 0) && _f) {
    this->set_exception(
                        std::make_exception_ptr(future_error(future_error_codes::broken_promise)));
}
```